src/dune_engine/build_system.ml

```diff
@@ -1790,6 +1793,94 @@ end = struct
     Memo.Build.return
       { deps; targets = Path.Build.Map.of_list_exn targets_digests }
 
+  let execute_rule_impl (rule : Rule.t) =
+    let* action, facts = exec_build_request rule.action.build in
+    execute_rule_impl_stage2 rule action ~observing_facts:facts
+      ~rule_kind:Normal_rule
+
+  let execute_action_generic ~observing_facts act ~capture_stdout =
+    let digest =
+      let
+          (* It is safe to ignore [dir] here as such actions are per-directory. *)
```

Comment:
As discussed, I think we need a way to "synchronize" when multiple code paths attempt to run the same action (and therefore, try to write to the same path).

It seems worth expanding here what "safe" means and why ignoring env, loc, alias is ok.

I think the property we want is that if the digest for two rules is equal then their outputs are equal, therefore it's OK to share them. In fact that's exactly the same property the `rule_digest` (coming from `compute_rule_digest`) tries to provide, so we should justify why we want a different digest here.

Reply:
Indeed, I'm updating this.

Reply:
Alright, I've added some documentation

Reply:
Looking back at this with a fresh eye, I think I'm still not doing the right thing here. I'm having another try

Reply:
alright, it should be good now.